sigstore/_verify.py

```diff
@@ -236,27 +246,91 @@ def verify(
 
         logger.debug("Successfully verified signature...")
 
-        # Retrieve the relevant Rekor entry to verify the inclusion proof and SET
-        entry = self._rekor.log.entries.retrieve.post(
-            signature.decode(),
-            sha256_artifact_hash,
-            base64.b64encode(certificate).decode(),
-        )
-        if entry is None:
-            return RekorEntryMissing(
-                signature=signature.decode(), sha256_artifact_hash=sha256_artifact_hash
+        entry: Optional[RekorEntry]
+        if offline_rekor_entry is not None:
+            # TODO(ww): This should all go in a separate API, probably under the
+            # RekorEntry class.
+            logger.debug("offline Rekor entry: checking consistency")
+
+            try:
+                entry_body = json.loads(base64.b64decode(offline_rekor_entry.body))
+            except Exception:
+                return VerificationFailure(
+                    reason="couldn't parse offline Rekor entry's body"
+                )
+
+            # The Rekor entry's body should be a hashedrekord object.
+            # TODO: This should use a real data model, ideally generated from
+            # Rekor's official JSON schema.
+            kind, version = entry_body.get("kind"), entry_body.get("apiVersion")
+            if kind != "hashedrekord" or version != "0.0.1":
+                return VerificationFailure(
+                    reason=(
+                        f"Rekor entry is of unsupported kind ('{kind}') or API "
+                        f"version ('{version}')"
+                    )
+                )
+
+            spec = entry_body["spec"]
+            expected_sig, expected_cert, expected_hash = (
+                spec["signature"]["content"],
+                load_pem_x509_certificate(
```

Comment:
This field can contain either a certificate, a certificate chain, or a public key, so we'll need to handle this. Not sure if there's other parts of the code that make this assumption also.

Reply:
I believe this is the only place we make that assumption, since the only other places we touch `hashedrekord` objects directly are during log entry submission and querying. 

In this case, it's okay for us to fail if we don't see a PEM-encoded certificate, right? This code only runs (currently) in the context of verification against a PEM-encoded certificate, so anything else would imply that the user has given us the wrong Rekor entry.

Reply:
(More generally, I think this demonstrates that I probably need to invert how I'm matching against the offline Rekor entry -- instead of unpacking the `body` and making sure it matches the cert/sig/etc, I should pack the latter into a `hashedrekord` payload and make sure it's strictly equal to the claimed entry.)

Reply:
So there's no support for verification by signing key rather than cert? (If not, feature request!)

Assuming only verification by cert is currently supported, I do think we should handle a certificate chain too.

Reply:
Correct! I'll file an issue for that.